apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicyElements.cs

```diff
@@ -0,0 +1,459 @@
+        {
+            /// <summary>
+            /// The default key element.
+            /// </summary>
+            public static Key Element { get; } = new Key();
+
+            internal Key() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "key";
+
+            /// <inheritdoc/>
+            object ISupportsExact<string>.ToPostPolicyValue(string value) => value;
+        }
+
+        /// <summary>
+        /// The allowed URL to which the client is redirected upon successful upload.
+        /// </summary>
+        internal sealed class SuccessActionRedirect : ISupportsExact<Uri>, ISupportsExact<string>, ISupportsStartsWith
+        {
+            /// <summary>
+            /// The default success-action-redirect element.
+            /// </summary>
+            public static SuccessActionRedirect Element { get; } = new SuccessActionRedirect();
+
+            internal SuccessActionRedirect() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "success-action-redirect";
+
+            /// <inheritdoc/>
+            object ISupportsExact<Uri>.ToPostPolicyValue(Uri value) => value.ToString();
+
+            /// <inheritdoc/>
+            object ISupportsExact<string>.ToPostPolicyValue(string value) => value;
+        }
+
+        /// <summary>
+        /// The status code returned to the client upon successful upload if 
+        /// <see cref="SuccessActionRedirect"/> is not specified.
+        /// Allowed values are 200, 201 and 204.
+        /// </summary>
+        internal sealed class SuccessActionStatus : ISupportsExact<HttpStatusCode>
+        {
+            /// <summary>
+            /// The default success-action-status element.
+            /// </summary>
+            public static SuccessActionStatus Element { get; } = new SuccessActionStatus();
+
+            internal SuccessActionStatus() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "success-action-status";
+
+            /// <inheritdoc/>
+            object ISupportsExact<HttpStatusCode>.ToPostPolicyValue(HttpStatusCode value) =>
+                ((int)value).ToString();
+        }
+
+        /// <summary>
+        /// Custom element used for requiering extra Google metadata to be included
+        /// in the post request.
+        /// </summary>
+        /// <typeparam name="TValue">The type for the values of the element.</typeparam>
+        internal sealed class GoogleMetadata<TValue> : ISupportsExact<TValue>, ISupportsStartsWith
+        {
+            private const string s_elementPrefix = "x-goog-meta-";
+
+            private readonly Func<TValue, object> _valueToObjectFunc;
+            private readonly string _elementName;
+
+            /// <summary>
+            /// Creates a new element with the given name.
+            /// If the given name is not prefixed with x-goog-meta- it will be so.
```

Comment:
This feels a little icky to me. Was it a documented requirement that all the other languages are doing? I think I'd prefer to say "we'll validate that it starts with x-goog-meta".